// @arch-ui/common/dist/common.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var theme = require("@arch-ui/theme");

const XL_DEVICE = 1199, LG_DEVICE = 991, MD_DEVICE = 768, SM_DEVICE = 575, smOnly = "@media (max-width: 575px)", mdOnly = "@media (min-width: 768px) and (max-width: 991px)", lgOnly = "@media (min-width: 991px) and (max-width: 1199px)", xlOnly = "@media (min-width: 1199px)", mdUp = "@media (min-width: 768px)", lgUp = "@media (min-width: 991px)", mdDown = "@media (max-width: 768px)", lgDown = "@media (max-width: 991px)", xlDown = "@media (max-width: 1199px)";

var mediaQueries = Object.freeze({
  __proto__: null,
  smOnly: smOnly,
  mdOnly: mdOnly,
  lgOnly: lgOnly,
  xlOnly: xlOnly,
  mdUp: mdUp,
  lgUp: lgUp,
  mdDown: mdDown,
  lgDown: lgDown,
  xlDown: xlDown
});

const uniformHeight = {
  appearance: "none",
  background: "none",
  border: "1px solid transparent",
  borderRadius: theme.borderRadius,
  boxSizing: "border-box",
  fontFamily: "inherit",
  fontSize: "1.0rem",
  margin: 0,
  minWidth: 1,
  padding: `${theme.gridSize}px ${1.5 * theme.gridSize}px`,
  transition: "box-shadow 100ms linear",
  verticalAlign: "middle",
  whiteSpace: "nowrap"
};

exports.mediaQueries = mediaQueries, exports.uniformHeight = uniformHeight;
