// @keystone-next/admin-ui/pages/HomePage/dist/admin-ui.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import { useState } from 'react';
import { jsx, Heading, Center, Inline, useTheme, VisuallyHidden } from '@keystone-ui/core';
import { makeDataGetter } from '@keystone-next/admin-ui-utils';
import { PlusIcon } from '@keystone-ui/icons/icons/PlusIcon';
import { DrawerController } from '@keystone-ui/modals';
import { LoadingDots } from '@keystone-ui/loading';
import { C as CreateItemDrawer } from '../../../dist/CreateItemDrawer-44bc7d45.esm.js';
import { P as PageContainer, H as HEADER_HEIGHT } from '../../../dist/GraphQLErrorNotice-80fc36a5.esm.js';
import { useQuery } from '@apollo/client';
import { useKeystone, useList } from '../../../context/dist/admin-ui.esm.js';
import { Link } from '../../../router/dist/admin-ui.esm.js';
import { useRouter } from 'next/router';
import '@babel/runtime/helpers/objectSpread2';
import 'fast-deep-equal';
import '@keystone-ui/toast';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@keystone-ui/button';
import '@keystone-ui/popover';
import '@keystone-ui/icons/icons/MoreHorizontalIcon';
import '@keystone-ui/icons/icons/ChevronRightIcon';
import '../../../dist/SignoutButton-be30cf42.esm.js';
import 'next/link';
import '@keystone-ui/notice';
import '@emotion/hash';
import '@keystone-next/types';
import '../../../dist/admin-meta-graphql-c0552487.esm.js';
import 'apollo-upload-client';

const ListCard = ({
  listKey,
  count
}) => {
  const {
    colors,
    palette,
    radii,
    spacing
  } = useTheme();
  const list = useList(listKey);
  const [isCreateModalOpen, setIsCreateModalOpen] = useState(false);
  const router = useRouter();
  return jsx("div", {
    css: {
      position: 'relative'
    }
  }, jsx(Link, {
    href: list.path,
    css: {
      backgroundColor: colors.background,
      borderColor: colors.border,
      borderRadius: radii.medium,
      borderWidth: 1,
      // boxShadow: shadow.s100,
      display: 'inline-block',
      minWidth: 280,
      padding: spacing.large,
      textDecoration: 'none',
      ':hover': {
        borderColor: palette.blue400
      },
      ':hover h3': {
        textDecoration: 'underline'
      }
    }
  }, jsx("h3", {
    css: {
      margin: `0 0 ${spacing.small}px 0`
    }
  }, list.label, " "), count.type === 'success' ? jsx("span", {
    css: {
      color: colors.foreground,
      textDecoration: 'none'
    }
  }, count.count, " item", count.count !== 1 ? 's' : '') : count.type === 'error' ? count.message : count.type === 'loading' ? jsx(LoadingDots, {
    label: `Loading count of ${list.plural}`,
    size: "small",
    tone: "passive"
  }) : 'No access'), jsx(CreateButton, {
    disabled: isCreateModalOpen,
    onClick: () => {
      setIsCreateModalOpen(true);
    }
  }, jsx(PlusIcon, {
    size: "large"
  }), jsx(VisuallyHidden, null, "Create ", list.singular)), jsx(DrawerController, {
    isOpen: isCreateModalOpen
  }, jsx(CreateItemDrawer, {
    listKey: list.key,
    onCreate: ({
      id
    }) => {
      router.push(`/${list.path}/${id}`);
    },
    onClose: () => {
      setIsCreateModalOpen(false);
    }
  })));
};

const CreateButton = props => {
  const theme = useTheme();
  return jsx("button", _extends({
    css: {
      alignItems: 'center',
      backgroundColor: theme.palette.neutral400,
      border: 0,
      borderRadius: theme.radii.xsmall,
      color: 'white',
      cursor: 'pointer',
      display: 'flex',
      height: 32,
      justifyContent: 'center',
      outline: 0,
      position: 'absolute',
      right: theme.spacing.large,
      top: theme.spacing.large,
      transition: 'background-color 80ms linear',
      width: 32,
      '&:hover, &:focus': {
        backgroundColor: theme.tones.positive.fill[0]
      }
    }
  }, props));
};

const HomePage = ({
  query
}) => {
  const {
    adminMeta: {
      lists
    },
    visibleLists
  } = useKeystone();
  let {
    data,
    error
  } = useQuery(query, {
    errorPolicy: 'all'
  });
  const dataGetter = makeDataGetter(data, error === null || error === void 0 ? void 0 : error.graphQLErrors);
  return jsx(PageContainer, {
    header: jsx(Heading, {
      type: "h3"
    }, "Dashboard")
  }, visibleLists.state === 'loading' ? jsx(Center, {
    css: {
      height: `calc(100vh - ${HEADER_HEIGHT}px)`
    }
  }, jsx(LoadingDots, {
    label: "Loading lists",
    size: "large",
    tone: "passive"
  })) : jsx(Inline, {
    gap: "large",
    paddingY: "xlarge"
  }, (() => {
    if (visibleLists.state === 'error') {
      return jsx("span", {
        css: {
          color: 'red'
        }
      }, visibleLists.error instanceof Error ? visibleLists.error.message : visibleLists.error[0].message);
    }

    return Object.keys(lists).map(key => {
      var _result$errors;

      if (!visibleLists.lists.has(key)) {
        return null;
      }

      const result = dataGetter.get(key); // TODO: Checking based on the message is bad, but we need to revisit GraphQL errors in
      // Keystone to fix it and that's a whole other can of worms...

      if (((_result$errors = result.errors) === null || _result$errors === void 0 ? void 0 : _result$errors[0].message) === 'You do not have access to this resource') {
        return jsx(ListCard, {
          count: {
            type: 'no-access'
          },
          key: key,
          listKey: key
        });
      }

      return jsx(ListCard, {
        count: data ? result.errors ? {
          type: 'error',
          message: result.errors[0].message
        } : {
          type: 'success',
          count: data[key].count
        } : {
          type: 'loading'
        },
        key: key,
        listKey: key
      });
    });
  })()));
};

export { HomePage };
