// @keystone-next/admin-ui/components/dist/admin-ui.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { useTheme, jsx } from '@keystone-ui/core';
export { C as CellLink } from '../../dist/CellLink-dd890306.esm.js';
export { E as ErrorBoundary, a as ErrorContainer } from '../../dist/Errors-ff56cb1b.esm.js';
export { G as GraphQLErrorNotice, L as Logo, N as Navigation, P as PageContainer } from '../../dist/GraphQLErrorNotice-80fc36a5.esm.js';
export { C as CreateItemDrawer } from '../../dist/CreateItemDrawer-44bc7d45.esm.js';
import '../../router/dist/admin-ui.esm.js';
import 'next/router';
import 'next/link';
import 'react';
import '@babel/runtime/helpers/defineProperty';
import '@keystone-ui/button';
import '@keystone-ui/icons/icons/AlertTriangleIcon';
import '@keystone-ui/popover';
import '@keystone-ui/icons/icons/MoreHorizontalIcon';
import '@keystone-ui/icons/icons/ChevronRightIcon';
import '../../dist/SignoutButton-be30cf42.esm.js';
import '@apollo/client';
import '../../context/dist/admin-ui.esm.js';
import '@keystone-ui/toast';
import '@keystone-ui/loading';
import '@keystone-ui/modals';
import '@babel/runtime/helpers/objectSpread2';
import '@emotion/hash';
import '@keystone-next/types';
import '../../dist/admin-meta-graphql-c0552487.esm.js';
import 'apollo-upload-client';
import '@keystone-next/admin-ui-utils';
import '@keystone-ui/notice';
import 'fast-deep-equal';

/**
 * This is the component you should use when you want the standard padding around a cell value
 */

const CellContainer = (_ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children"]);

  const {
    spacing
  } = useTheme();
  return jsx("div", _extends({
    css: {
      padding: spacing.small
    }
  }, props), children);
};

export { CellContainer };
