// @keystonejs/fields-content/views/editor/blocks/unordered-list/dist/fields-content.esm.js
import '@babel/runtime/helpers/objectWithoutProperties';
import '@emotion/core';
import { createElement } from 'react';
import '@arch-ui/theme';
import '@babel/runtime/helpers/extends';
import { type as type$1 } from '../../paragraph/dist/fields-content.esm.js';
import { ListUnorderedIcon } from '@primer/octicons-react';
import '@arch-ui/color-utils';
import '@arch-ui/tooltip';
import '@arch-ui/typography';
import { T as ToolbarButton } from '../../../../../dist/toolbar-components-c9e1fefe.esm.js';
import { h as hasAncestorBlock, a as hasBlock } from '../../../../../dist/utils-d86d79bf.esm.js';
import { type as type$2 } from '../../list-item/dist/fields-content.esm.js';

let handleListButtonClick = (editor, editorState, type) => {
  let isListItem = hasBlock(editorState, type$2);
  let isOrderedList = hasAncestorBlock(editorState, type);
  let otherListType = type === 'ordered-list' ? 'unordered-list' : 'ordered-list';

  if (isListItem && isOrderedList) {
    editor.setBlocks(type$1);
    editor.unwrapBlock(type);
  } else if (isListItem) {
    editor.unwrapBlock(otherListType);
    editor.wrapBlock(type);
  } else {
    editor.setBlocks(type$2).wrapBlock(type);
  }

  editor.focus();
};

let type = 'unordered-list';
function ToolbarElement({
  editor,
  editorState
}) {
  return /*#__PURE__*/createElement(ToolbarButton, {
    label: "Unordered List",
    icon: /*#__PURE__*/createElement(ListUnorderedIcon, null),
    isActive: hasAncestorBlock(editorState, type),
    onClick: () => {
      handleListButtonClick(editor, editorState, type);
    }
  });
}
let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter in an empty list item,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && hasAncestorBlock(editor.value, type) && editor.value.focusText.text === '') {
      editor.setBlocks(type$1).unwrapBlock(type);
      return;
    }

    next();
  }

}];
function Node({
  attributes,
  children
}) {
  return /*#__PURE__*/createElement("ul", attributes, children);
}
const getSchema = () => ({
  nodes: [{
    match: {
      type: type$2
    },
    min: 0
  }],

  normalize(editor, error) {
    switch (error.code) {
      case 'child_type_invalid':
        {
          editor.unwrapBlockByKey(error.node.key, type);
          return;
        }
    }
  }

});

export { Node, ToolbarElement, getPlugins, getSchema, type };
