// @keystone-next/admin-ui/pages/ItemPage/dist/admin-ui.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var copyToClipboard = require('clipboard-copy');
var Link = require('next/link');
var router = require('next/router');
var React = require('react');
var button = require('@keystone-ui/button');
var core = require('@keystone-ui/core');
var loading = require('@keystone-ui/loading');
var ClipboardIcon = require('@keystone-ui/icons/icons/ClipboardIcon');
var ChevronRightIcon = require('@keystone-ui/icons/icons/ChevronRightIcon');
var modals = require('@keystone-ui/modals');
var notice = require('@keystone-ui/notice');
var toast = require('@keystone-ui/toast');
var tooltip = require('@keystone-ui/tooltip');
var fields = require('@keystone-ui/fields');
var adminUiUtils = require('@keystone-next/admin-ui-utils');
var client = require('@apollo/client');
var context_dist_adminUi = require('../../../context/dist/admin-ui.cjs.prod.js');
var GraphQLErrorNotice = require('../../../dist/GraphQLErrorNotice-b55ab5a7.cjs.prod.js');
require('@babel/runtime/helpers/objectSpread2');
require('fast-deep-equal');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
require('@emotion/hash');
require('@keystone-next/types');
require('../../../dist/admin-meta-graphql-5b37ade6.cjs.prod.js');
require('apollo-upload-client');
require('@keystone-ui/popover');
require('@keystone-ui/icons/icons/MoreHorizontalIcon');
require('../../../dist/SignoutButton-a435a427.cjs.prod.js');
require('../../../router/dist/admin-ui.cjs.prod.js');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var copyToClipboard__default = /*#__PURE__*/_interopDefault(copyToClipboard);
var Link__default = /*#__PURE__*/_interopDefault(Link);

/**
 * NOTE: should probably come from the DS?
 */

const Container = (_ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children"]);

  return core.jsx("div", _extends({
    css: {
      minWidth: 0,
      // fix flex text truncation
      maxWidth: 1080 // marginLeft: 'auto',
      // marginRight: 'auto',

    }
  }, props), children);
};

function useEventCallback(callback) {
  const callbackRef = React.useRef(callback);
  const cb = React.useCallback((...args) => {
    return callbackRef.current(...args);
  }, []);
  React.useEffect(() => {
    callbackRef.current = callback;
  });
  return cb;
}

function ItemForm({
  listKey,
  itemGetter,
  selectedFields,
  fieldModes,
  showDelete
}) {
  var _useMemo, _itemGetter$data4, _itemGetter$data5;

  const list = context_dist_adminUi.useList(listKey);
  const [update, {
    loading,
    error,
    data
  }] = client.useMutation(client.gql`mutation ($data: ${list.gqlNames.updateInputName}!, $id: ID!) {
      item: ${list.gqlNames.updateMutationName}(id: $id, data: $data) {
        ${selectedFields}
      }
    }`, {
    errorPolicy: 'all'
  });
  itemGetter = (_useMemo = React.useMemo(() => {
    if (data) {
      return adminUiUtils.makeDataGetter(data, error === null || error === void 0 ? void 0 : error.graphQLErrors).get('item');
    }
  }, [data, error])) !== null && _useMemo !== void 0 ? _useMemo : itemGetter;
  const [state, setValue] = React.useState(() => {
    const value = adminUiUtils.deserializeValue(list.fields, itemGetter);
    return {
      value,
      item: itemGetter.data
    };
  });

  if (!loading && state.item !== itemGetter.data && (itemGetter.errors || []).every(x => {
    var _x$path;

    return ((_x$path = x.path) === null || _x$path === void 0 ? void 0 : _x$path.length) !== 1;
  })) {
    const value = adminUiUtils.deserializeValue(list.fields, itemGetter);
    setValue({
      value,
      item: itemGetter.data
    });
  }

  const {
    changedFields,
    dataForUpdate
  } = adminUiUtils.useChangedFieldsAndDataForUpdate(list.fields, itemGetter, state.value);
  const invalidFields = adminUiUtils.useInvalidFields(list.fields, state.value);
  const [forceValidation, setForceValidation] = React.useState(false);
  const toasts = toast.useToasts();
  const onSave = useEventCallback(() => {
    const newForceValidation = invalidFields.size !== 0;
    setForceValidation(newForceValidation);
    if (newForceValidation) return;
    update({
      variables: {
        data: dataForUpdate,
        id: itemGetter.get('id').data
      }
    }) // TODO -- Experimenting with less detail in the toasts, so the data lines are commented
    // out below. If we're happy with this, clean up the unused lines.
    .then(({
      /* data, */
      errors
    }) => {
      // we're checking for path.length === 1 because errors with a path larger than 1 will
      // be field level errors which are handled seperately and do not indicate a failure to
      // update the item
      const error = errors === null || errors === void 0 ? void 0 : errors.find(x => {
        var _x$path2;

        return ((_x$path2 = x.path) === null || _x$path2 === void 0 ? void 0 : _x$path2.length) === 1;
      });

      if (error) {
        toasts.addToast({
          title: 'Failed to update item',
          tone: 'negative',
          message: error.message
        });
      } else {
        toasts.addToast({
          // title: data.item[list.labelField] || data.item.id,
          tone: 'positive',
          title: 'Saved successfully' // message: 'Saved successfully',

        });
      }
    }).catch(err => {
      toasts.addToast({
        title: 'Failed to update item',
        tone: 'negative',
        message: err.message
      });
    });
  });
  return core.jsx(core.Box, {
    marginTop: "xlarge"
  }, core.jsx(GraphQLErrorNotice.GraphQLErrorNotice, {
    networkError: error === null || error === void 0 ? void 0 : error.networkError // we're checking for path.length === 1 because errors with a path larger than 1 will be field level errors
    // which are handled seperately and do not indicate a failure to update the item
    ,
    errors: error === null || error === void 0 ? void 0 : error.graphQLErrors.filter(x => {
      var _x$path3;

      return ((_x$path3 = x.path) === null || _x$path3 === void 0 ? void 0 : _x$path3.length) === 1;
    })
  }), core.jsx(adminUiUtils.Fields, {
    fieldModes: fieldModes,
    fields: list.fields,
    forceValidation: forceValidation,
    invalidFields: invalidFields,
    onChange: React.useCallback(value => {
      setValue(state => ({
        item: state.item,
        value: value(state.value)
      }));
    }, [setValue]),
    value: state.value
  }), core.jsx(Toolbar, {
    onSave: onSave,
    hasChangedFields: !!changedFields.size,
    onReset: useEventCallback(() => {
      setValue({
        item: itemGetter.data,
        value: adminUiUtils.deserializeValue(list.fields, itemGetter)
      });
    }),
    loading: loading,
    deleteButton: React.useMemo(() => {
      var _itemGetter$data$list, _itemGetter$data, _itemGetter$data2, _itemGetter$data3;

      return showDelete ? core.jsx(DeleteButton, {
        list: list,
        itemLabel: (_itemGetter$data$list = (_itemGetter$data = itemGetter.data) === null || _itemGetter$data === void 0 ? void 0 : _itemGetter$data[list.labelField]) !== null && _itemGetter$data$list !== void 0 ? _itemGetter$data$list : (_itemGetter$data2 = itemGetter.data) === null || _itemGetter$data2 === void 0 ? void 0 : _itemGetter$data2.id,
        itemId: (_itemGetter$data3 = itemGetter.data) === null || _itemGetter$data3 === void 0 ? void 0 : _itemGetter$data3.id
      }) : undefined;
    }, [showDelete, list, (_itemGetter$data4 = itemGetter.data) === null || _itemGetter$data4 === void 0 ? void 0 : _itemGetter$data4[list.labelField], (_itemGetter$data5 = itemGetter.data) === null || _itemGetter$data5 === void 0 ? void 0 : _itemGetter$data5.id])
  }));
}

function DeleteButton({
  itemLabel,
  itemId,
  list
}) {
  const toasts = toast.useToasts();
  const [deleteItem, {
    loading
  }] = client.useMutation(client.gql`mutation ($id: ID!) {
      ${list.gqlNames.deleteMutationName}(id: $id) {
        id
      }
    }`, {
    variables: {
      id: itemId
    }
  });
  const [isOpen, setIsOpen] = React.useState(false);
  const router$1 = router.useRouter();
  return core.jsx(React.Fragment, null, core.jsx(button.Button, {
    tone: "negative",
    onClick: () => {
      setIsOpen(true);
    }
  }, "Delete"), core.jsx(modals.AlertDialog // TODO: change the copy in the title and body of the modal
  , {
    title: "Delete Confirmation",
    isOpen: isOpen,
    tone: "negative",
    actions: {
      confirm: {
        label: 'Delete',
        action: async () => {
          await deleteItem().catch(err => {
            toasts.addToast({
              title: 'Failed to delete item',
              message: err.message,
              tone: 'negative'
            });
          });
          router$1.push(`/${list.path}`);
          toasts.addToast({
            title: itemLabel,
            message: 'Deleted successfully',
            tone: 'positive'
          });
        },
        loading
      },
      cancel: {
        label: 'Cancel',
        action: () => {
          setIsOpen(false);
        }
      }
    }
  }, "Are you sure you want to delete ", core.jsx("strong", null, itemLabel), "?"));
}

const ItemPage = ({
  listKey
}) => {
  var _dataGetter$data2, _dataGetter$data2$key, _dataGetter$data2$key2, _dataGetter$data2$key3;

  const router$1 = router.useRouter();
  const {
    id
  } = router$1.query;
  const list = context_dist_adminUi.useList(listKey);
  const {
    palette,
    spacing,
    typography
  } = core.useTheme();
  const {
    query,
    selectedFields
  } = React.useMemo(() => {
    let selectedFields = Object.keys(list.fields).map(fieldPath => {
      return list.fields[fieldPath].controller.graphqlSelection;
    }).join('\n');
    return {
      selectedFields,
      query: client.gql`
        query ItemPage($id: ID!, $listKey: String!) {
          item: ${list.gqlNames.itemQueryName}(where: {id: $id}) {
            ${selectedFields}
          }
          keystone {
            adminMeta {
              list(key: $listKey) {
                hideCreate
                hideDelete
                fields {
                  path
                  itemView(id: $id) {
                    fieldMode
                  }
                }
              }
            }
          }
        }
      `
    };
  }, [list]);
  let {
    data,
    error,
    loading: loading$1
  } = client.useQuery(query, {
    variables: {
      id,
      listKey
    },
    errorPolicy: 'all',
    skip: id === undefined
  });
  loading$1 || (loading$1 = id === undefined);
  const dataGetter = adminUiUtils.makeDataGetter(data, error === null || error === void 0 ? void 0 : error.graphQLErrors);
  let itemViewFieldModesByField = React.useMemo(() => {
    var _dataGetter$data, _dataGetter$data$keys, _dataGetter$data$keys2, _dataGetter$data$keys3, _dataGetter$data$keys4;

    let itemViewFieldModesByField = {};
    (_dataGetter$data = dataGetter.data) === null || _dataGetter$data === void 0 ? void 0 : (_dataGetter$data$keys = _dataGetter$data.keystone) === null || _dataGetter$data$keys === void 0 ? void 0 : (_dataGetter$data$keys2 = _dataGetter$data$keys.adminMeta) === null || _dataGetter$data$keys2 === void 0 ? void 0 : (_dataGetter$data$keys3 = _dataGetter$data$keys2.list) === null || _dataGetter$data$keys3 === void 0 ? void 0 : (_dataGetter$data$keys4 = _dataGetter$data$keys3.fields) === null || _dataGetter$data$keys4 === void 0 ? void 0 : _dataGetter$data$keys4.forEach(field => {
      var _field$itemView;

      if (field !== null && field.path !== null && (field === null || field === void 0 ? void 0 : (_field$itemView = field.itemView) === null || _field$itemView === void 0 ? void 0 : _field$itemView.fieldMode) != null) {
        itemViewFieldModesByField[field.path] = field.itemView.fieldMode;
      }
    });
    return itemViewFieldModesByField;
  }, [(_dataGetter$data2 = dataGetter.data) === null || _dataGetter$data2 === void 0 ? void 0 : (_dataGetter$data2$key = _dataGetter$data2.keystone) === null || _dataGetter$data2$key === void 0 ? void 0 : (_dataGetter$data2$key2 = _dataGetter$data2$key.adminMeta) === null || _dataGetter$data2$key2 === void 0 ? void 0 : (_dataGetter$data2$key3 = _dataGetter$data2$key2.list) === null || _dataGetter$data2$key3 === void 0 ? void 0 : _dataGetter$data2$key3.fields]);
  const metaQueryErrors = dataGetter.get('keystone').errors; // NOTE: The create button is always hidden on this page for now, while we work on the
  // placment of the save and delete buttons.

  const hideCreate = true; // data?.keystone.adminMeta.list.hideCreate;

  return core.jsx(GraphQLErrorNotice.PageContainer, {
    header: core.jsx(Container, {
      css: {
        alignItems: 'center',
        display: 'flex',
        flex: 1,
        justifyContent: 'space-between'
      }
    }, core.jsx("div", {
      css: {
        alignItems: 'center',
        display: 'flex',
        flex: 1,
        minWidth: 0
      }
    }, core.jsx(core.Heading, {
      type: "h3"
    }, core.jsx(Link__default['default'], {
      href: `/${list.path}`,
      passHref: true
    }, core.jsx("a", {
      css: {
        textDecoration: 'none'
      }
    }, list.label))), core.jsx("div", {
      css: {
        color: palette.neutral500,
        marginLeft: spacing.xsmall,
        marginRight: spacing.xsmall
      }
    }, core.jsx(ChevronRightIcon.ChevronRightIcon, null)), core.jsx(core.Heading, {
      as: "h1",
      type: "h3",
      css: {
        minWidth: 0,
        maxWidth: '100%',
        overflow: 'hidden',
        flex: 1,
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap'
      }
    }, loading$1 ? 'Loading...' : data && data.item && (data.item[list.labelField] || data.item.id) || id)), !hideCreate )
  }, loading$1 ? core.jsx(core.Center, {
    css: {
      height: `calc(100vh - ${GraphQLErrorNotice.HEADER_HEIGHT}px)`
    }
  }, core.jsx(loading.LoadingDots, {
    label: "Loading item data",
    size: "large",
    tone: "passive"
  })) : metaQueryErrors ? core.jsx(core.Box, {
    marginY: "xlarge"
  }, core.jsx(notice.Notice, {
    tone: "negative"
  }, metaQueryErrors[0].message)) : core.jsx(React.Fragment, null, core.jsx(ColumnLayout, null, core.jsx(ItemForm, {
    fieldModes: itemViewFieldModesByField,
    selectedFields: selectedFields,
    showDelete: !data.keystone.adminMeta.list.hideDelete,
    listKey: listKey,
    itemGetter: dataGetter.get('item')
  }), core.jsx(StickySidebar, null, core.jsx(fields.FieldLabel, null, "Item ID"), core.jsx("div", {
    css: {
      display: 'flex',
      alignItems: 'center'
    }
  }, core.jsx(fields.TextInput, {
    css: {
      marginRight: spacing.medium,
      fontFamily: typography.fontFamily.monospace,
      fontSize: typography.fontSize.small
    },
    readOnly: true,
    value: data.item.id
  }), core.jsx(tooltip.Tooltip, {
    content: "Copy ID"
  }, props => core.jsx(button.Button, _extends({}, props, {
    "aria-label": "Copy ID",
    onClick: () => {
      copyToClipboard__default['default'](data.item.id);
    }
  }), core.jsx(ClipboardIcon.ClipboardIcon, {
    size: "small"
  }))))))));
};
// ------------------------------


const Toolbar = /*#__PURE__*/React.memo(function Toolbar({
  hasChangedFields,
  loading,
  onSave,
  onReset,
  deleteButton
}) {
  const {
    colors,
    spacing
  } = core.useTheme();
  return core.jsx("div", {
    css: {
      background: colors.background,
      borderTop: `1px solid ${colors.border}`,
      bottom: 0,
      display: 'flex',
      justifyContent: 'space-between',
      marginTop: spacing.xlarge,
      paddingBottom: spacing.xlarge,
      paddingTop: spacing.xlarge,
      position: 'sticky',
      zIndex: 10
    }
  }, core.jsx(core.Stack, {
    align: "center",
    across: true,
    gap: "small"
  }, core.jsx(button.Button, {
    isDisabled: !hasChangedFields,
    isLoading: loading,
    weight: "bold",
    tone: "active",
    onClick: onSave
  }, "Save changes"), hasChangedFields ? core.jsx(button.Button, {
    weight: "none",
    onClick: onReset
  }, "Reset changes") : core.jsx(core.Text, {
    weight: "medium",
    paddingX: "large",
    color: "neutral600"
  }, "No changes")), deleteButton);
});

const ColumnLayout = props => {
  const {
    spacing
  } = core.useTheme();
  return core.jsx(Container, {
    css: {
      position: 'relative'
    }
  }, core.jsx("div", _extends({
    css: {
      alignItems: 'start',
      display: 'grid',
      gap: spacing.xlarge,
      gridTemplateColumns: `2fr 1fr`
    }
  }, props)));
};

const StickySidebar = props => {
  const {
    spacing
  } = core.useTheme();
  return core.jsx("div", _extends({
    css: {
      marginTop: spacing.xlarge,
      marginBottom: spacing.xxlarge,
      position: 'sticky',
      top: spacing.xlarge
    }
  }, props));
};

exports.ItemPage = ItemPage;
