// @arch-ui/alert/dist/alert.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var styled = require("@emotion/styled"), theme = require("@arch-ui/theme");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

var styled__default = _interopDefault(styled);

const boldBackgroundColor = {
  info: theme.colors.primary,
  success: theme.colors.create,
  warning: theme.colors.warning,
  danger: theme.colors.danger
}, boldTextColor = {
  info: "white",
  success: "white",
  warning: "white",
  danger: "white"
}, subtleBackgroundColor = {
  info: theme.colors.B.L80,
  success: theme.colors.G.L80,
  warning: theme.colors.Y.L80,
  danger: theme.colors.R.L80
}, subtleTextColor = {
  info: theme.colors.B.D20,
  success: theme.colors.G.D20,
  warning: theme.colors.Y.D20,
  danger: theme.colors.R.D20
}, Alert = styled__default.default.div(({appearance: appearance = "info", variant: variant = "subtle"}) => ({
  backgroundColor: "bold" === variant ? boldBackgroundColor[appearance] : subtleBackgroundColor[appearance],
  color: "bold" === variant ? boldTextColor[appearance] : subtleTextColor[appearance],
  borderRadius: theme.borderRadius,
  display: "flex",
  fontWeight: "bold" === variant ? 500 : null,
  minWidth: 1,
  padding: "0.9em 1.2em",
  margin: "0.9em 0",
  "& a": {
    color: "bold" === variant ? boldTextColor[appearance] : subtleTextColor[appearance],
    textDecoration: "underline"
  }
}));

exports.Alert = Alert;
