// @keystone-next/admin-ui/dist/GraphQLErrorNotice-b55ab5a7.cjs.prod.js
'use strict';

var _extends = require('@babel/runtime/helpers/extends');
var core = require('@keystone-ui/core');
var React = require('react');
var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var router = require('next/router');
var button = require('@keystone-ui/button');
var popover = require('@keystone-ui/popover');
var MoreHorizontalIcon = require('@keystone-ui/icons/icons/MoreHorizontalIcon');
var ChevronRightIcon = require('@keystone-ui/icons/icons/ChevronRightIcon');
var SignoutButton = require('./SignoutButton-a435a427.cjs.prod.js');
var context_dist_adminUi = require('../context/dist/admin-ui.cjs.prod.js');
var router_dist_adminUi = require('../router/dist/admin-ui.cjs.prod.js');
var Link = require('next/link');
var notice = require('@keystone-ui/notice');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var Link__default = /*#__PURE__*/_interopDefault(Link);

/* @jsx jsx */
const Logo = () => {
  var _adminConfig$componen;

  const {
    headingStyles
  } = core.useTheme();
  const {
    adminConfig
  } = context_dist_adminUi.useKeystone();

  if ((_adminConfig$componen = adminConfig.components) !== null && _adminConfig$componen !== void 0 && _adminConfig$componen.Logo) {
    return core.jsx(adminConfig.components.Logo, null);
  }

  return core.jsx(core.H3, null, core.jsx(Link__default['default'], {
    href: "/",
    passHref: true
  }, core.jsx("a", {
    css: {
      color: headingStyles.h3.color,
      textDecoration: 'none'
    }
  }, "KeystoneJS")));
};

const NavItem = ({
  href,
  children
}) => {
  const {
    palette,
    spacing,
    radii,
    typography
  } = core.useTheme();
  const router$1 = router.useRouter();
  const isSelected = router$1.pathname === href || router$1.pathname.split('/')[1] === href.split('/')[1];
  return core.jsx(router_dist_adminUi.Link, {
    "aria-current": isSelected ? 'location' : false,
    href: href,
    css: {
      background: 'transparent',
      borderBottomRightRadius: radii.xsmall,
      borderTopRightRadius: radii.xsmall,
      color: palette.neutral700,
      display: 'block',
      fontWeight: typography.fontWeight.medium,
      marginBottom: spacing.xsmall,
      marginRight: spacing.xlarge,
      padding: `${spacing.small}px ${spacing.xlarge}px`,
      position: 'relative',
      textDecoration: 'none',
      ':hover': {
        background: palette.blue50,
        color: palette.blue500
      },
      '&[aria-current=location]': {
        background: palette.neutral200,
        color: palette.neutral900
      }
    }
  }, children);
};

const AuthenticatedItem = ({
  item
}) => {
  const {
    spacing,
    typography
  } = core.useTheme();
  return core.jsx("div", {
    css: {
      alignItems: 'center',
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      padding: spacing.xlarge
    }
  }, core.jsx("div", {
    css: {
      fontSize: typography.fontSize.small
    }
  }, "Signed in as ", core.jsx("strong", null, item.label)), core.jsx(popover.Popover, {
    placement: "bottom",
    triggerRenderer: ({
      triggerProps
    }) => core.jsx(button.Button, _extends({
      size: "small",
      style: {
        padding: 0,
        width: 36
      },
      "aria-label": "Links and signout"
    }, triggerProps), core.jsx(MoreHorizontalIcon.MoreHorizontalIcon, null))
  }, core.jsx(core.Stack, {
    gap: "medium",
    padding: "large",
    dividers: "between"
  }, core.jsx(PopoverLink, {
    target: "_blank",
    href: "/api/graphql"
  }, "API Explorer"), core.jsx(PopoverLink, {
    target: "_blank",
    href: "https://github.com/keystonejs/keystone"
  }, "GitHub Repository"), core.jsx(PopoverLink, {
    target: "_blank",
    href: "https://www.keystonejs.com/documentation"
  }, "Keystone Documentation"), core.jsx(SignoutButton.SignoutButton, null))));
};

const PopoverLink = (_ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children"]);

  const {
    typography
  } = core.useTheme();
  return core.jsx("a", _extends({
    css: {
      alignItems: 'center',
      display: 'flex',
      fontSize: typography.fontSize.small,
      textDecoration: 'none'
    }
  }, props), children, core.jsx(ChevronRightIcon.ChevronRightIcon, {
    size: "small"
  }));
};

const Navigation = () => {
  const {
    adminMeta: {
      lists
    },
    authenticatedItem,
    visibleLists
  } = context_dist_adminUi.useKeystone();
  return core.jsx("div", {
    css: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center'
    }
  }, authenticatedItem.state === 'authenticated' && core.jsx(AuthenticatedItem, {
    item: authenticatedItem
  }), core.jsx("nav", null, core.jsx(NavItem, {
    href: "/"
  }, "Dashboard"), (() => {
    if (visibleLists.state === 'loading') return null;

    if (visibleLists.state === 'error') {
      return core.jsx("span", {
        css: {
          color: 'red'
        }
      }, visibleLists.error instanceof Error ? visibleLists.error.message : visibleLists.error[0].message);
    }

    return Object.keys(lists).map(key => {
      if (!visibleLists.lists.has(key)) {
        return null;
      }

      const list = lists[key];
      return core.jsx(NavItem, {
        key: key,
        href: `/${list.path}`
      }, lists[key].label);
    });
  })()));
};

const HEADER_HEIGHT = 80;

const PageWrapper = props => {
  // const { colors } = useTheme();
  return core.jsx(React.Fragment, null, core.jsx("style", null, `body { overflow: hidden; }`), core.jsx("div", _extends({
    css: {
      // background: colors.background,
      display: 'grid',
      gridTemplateColumns: `minmax(300px, 1fr) 4fr`,
      gridTemplateRows: `${HEADER_HEIGHT}px auto`,
      height: '100vh'
    }
  }, props)));
};

const Sidebar = props => {
  // const { colors } = useTheme();
  return core.jsx("aside", _extends({
    css: {
      // borderRight: `1px solid ${colors.border}`,
      minWidth: 0,
      // resolves collapsing issues in children
      overflowY: 'auto',
      WebkitOverflowScrolling: 'touch'
    }
  }, props));
};

const Content = props => {
  const {
    colors,
    spacing
  } = core.useTheme();
  return core.jsx("main", _extends({
    css: {
      backgroundColor: colors.background,
      boxSizing: 'border-box',
      minWidth: 0,
      // resolves collapsing issues in children
      paddingLeft: spacing.xlarge,
      paddingRight: spacing.xlarge,
      overflowY: 'auto',
      WebkitOverflowScrolling: 'touch'
    }
  }, props));
};

const PageContainer = ({
  children,
  header
}) => {
  const {
    colors,
    spacing
  } = core.useTheme();
  return core.jsx(PageWrapper, null, core.jsx("div", {
    css: {
      alignItems: 'center',
      // borderRight: `1px solid ${colors.border}`,
      borderBottom: `1px solid ${colors.border}`,
      display: 'flex',
      justifyContent: 'space-between',
      paddingLeft: spacing.xlarge,
      paddingRight: spacing.xlarge
    }
  }, core.jsx(Logo, null)), core.jsx("header", {
    css: {
      alignItems: 'center',
      backgroundColor: colors.background,
      borderBottom: `1px solid ${colors.border}`,
      display: 'flex',
      justifyContent: 'space-between',
      minWidth: 0,
      // fix flex text truncation
      paddingLeft: spacing.xlarge,
      paddingRight: spacing.xlarge
    }
  }, header), core.jsx(Sidebar, null, core.jsx(Navigation, null)), core.jsx(Content, null, children));
};

function GraphQLErrorNotice({
  errors,
  networkError
}) {
  if (networkError) {
    return /*#__PURE__*/React__default['default'].createElement(notice.Notice, {
      tone: "negative"
    }, networkError.message);
  }

  if (errors !== null && errors !== void 0 && errors.length) {
    return /*#__PURE__*/React__default['default'].createElement(core.Stack, {
      gap: "small"
    }, errors.map(err => {
      var _errButAny$data, _errButAny$data$messa;

      const errButAny = err;

      if (err.name === 'ValidationFailureError' && (_errButAny$data = errButAny.data) !== null && _errButAny$data !== void 0 && (_errButAny$data$messa = _errButAny$data.messages) !== null && _errButAny$data$messa !== void 0 && _errButAny$data$messa.length) {
        return errButAny.data.messages.map(message => /*#__PURE__*/React__default['default'].createElement(notice.Notice, {
          tone: "negative"
        }, message));
      }

      return /*#__PURE__*/React__default['default'].createElement(notice.Notice, {
        tone: "negative"
      }, err.message);
    }));
  }

  return null;
}

exports.GraphQLErrorNotice = GraphQLErrorNotice;
exports.HEADER_HEIGHT = HEADER_HEIGHT;
exports.Logo = Logo;
exports.Navigation = Navigation;
exports.PageContainer = PageContainer;
