// @arch-ui/fields/dist/fields.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _objectWithoutProperties = require("@babel/runtime/helpers/objectWithoutProperties"), _extends = require("@babel/runtime/helpers/extends"), core = require("@emotion/core"), theme = require("@arch-ui/theme"), octiconsReact = require("@primer/octicons-react"), lozenge = require("@arch-ui/lozenge");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

var _objectWithoutProperties__default = _interopDefault(_objectWithoutProperties), _extends__default = _interopDefault(_extends);

const FieldContainer = props => core.jsx("div", _extends__default.default({
  "data-selector": "field-container",
  css: {
    marginBottom: 2 * theme.gridSize
  }
}, props)), FieldLabel = _ref => {
  let {errors: errors = [], field: field} = _ref, props = _objectWithoutProperties__default.default(_ref, [ "errors", "field" ]);
  const accessError = errors.find(error => error instanceof Error && "AccessDeniedError" === error.name);
  return core.jsx("label", _extends__default.default({
    css: {
      color: theme.colors.N60,
      fontSize: "0.9rem",
      fontWeight: 500,
      paddingBottom: theme.gridSize,
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between"
    }
  }, props), field.label, core.jsx("span", null, field.isRequired && core.jsx(lozenge.Lozenge, {
    appearance: "primary"
  }, "Required"), accessError && core.jsx(octiconsReact.ShieldIcon, {
    title: accessError.message,
    css: {
      color: theme.colors.N20,
      margin: "0 1em"
    }
  })));
}, FieldDescription = _ref2 => {
  let {text: text} = _ref2, props = _objectWithoutProperties__default.default(_ref2, [ "text" ]);
  return text ? core.jsx("p", _extends__default.default({
    css: {
      margin: "0 0 8px",
      color: theme.colors.N60,
      fontSize: "0.9rem"
    }
  }, props), text) : null;
}, FieldInput = props => core.jsx("div", _extends__default.default({
  css: {
    display: "flex"
  }
}, props)), Currency = props => core.jsx("span", _extends__default.default({
  css: {
    alignContent: "center",
    backgroundColor: theme.colors.N10,
    border: "1px solid transparent",
    borderBottomRightRadius: 0,
    borderColor: theme.colors.N20,
    borderRadius: 4,
    borderRight: 0,
    borderTopRightRadius: 0,
    display: "flex",
    fontSize: "0.9rem",
    marginRight: -2,
    padding: 10
  }
}, props));

exports.Currency = Currency, exports.FieldContainer = FieldContainer, exports.FieldDescription = FieldDescription, 
exports.FieldInput = FieldInput, exports.FieldLabel = FieldLabel;
