// @keystonejs/fields-content/dist/toolbar-components-5704db01.cjs.dev.js
'use strict';

var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var core = require('@emotion/core');
var theme = require('@arch-ui/theme');
var _extends = require('@babel/runtime/helpers/extends');
var colorUtils = require('@arch-ui/color-utils');
var Tooltip = require('@arch-ui/tooltip');
var typography = require('@arch-ui/typography');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _objectWithoutProperties__default = /*#__PURE__*/_interopDefault(_objectWithoutProperties);
var _extends__default = /*#__PURE__*/_interopDefault(_extends);
var Tooltip__default = /*#__PURE__*/_interopDefault(Tooltip);

let ToolbarButton = (_ref) => {
  let {
    isActive,
    label,
    icon,
    as: Tag = 'button',
    tooltipPlacement = 'top'
  } = _ref,
      props = _objectWithoutProperties__default['default'](_ref, ["isActive", "label", "icon", "as", "tooltipPlacement"]);

  return core.jsx(Tooltip__default['default'], {
    placement: tooltipPlacement,
    css: {
      margin: theme.gridSize * 2
    },
    content: label
  }, ref => core.jsx(Tag, _extends__default['default']({}, Tag === 'button' ? {
    type: 'button'
  } : null, {
    css: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: 'transparent',
      border: 0,
      cursor: 'pointer',
      color: isActive ? theme.colors.primary : 'white',
      ':hover,:focus': {
        color: isActive ? colorUtils.darken(theme.colors.primary, 10) : colorUtils.lighten(theme.colors.primary, 40)
      },
      ':active': {
        color: isActive ? colorUtils.darken(theme.colors.primary, 25) : colorUtils.lighten(theme.colors.primary, 10)
      },
      fontSize: 16,
      outline: 'none'
    },
    ref: ref
  }, props), icon, core.jsx(typography.A11yText, null, label)));
};

exports.ToolbarButton = ToolbarButton;
