// @keystonejs/fields-content/dist/unordered-list-39431930.cjs.dev.js
'use strict';

var path = require('path');
var Block_dist_fieldsContent = require('../Block/dist/fields-content.cjs.dev.js');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var path__default = /*#__PURE__*/_interopDefault(path);

const pkgDir = path__default['default'].dirname(require.resolve('@keystonejs/fields-content/package.json'));
const resolveView = pathname => path__default['default'].join(pkgDir, pathname);

class BlockquoteBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'blockquote';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/blockquote')];
  }

}

class CaptionBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'caption';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/caption')];
  }

}

class HeadingBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'heading';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/heading'), ...new ParagraphBlock().getAdminViews()];
  }

}

class ImageBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'image';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/image')];
  }

}

class ImageContainerBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'image-container';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/image-container'), ...new ImageBlock().getAdminViews(), ...new CaptionBlock().getAdminViews()];
  }

}

class LinkBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'link';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/link')];
  }

}

class ListItemBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'list-item';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/list-item')];
  }

}

class OrderedListBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'ordered-list';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/ordered-list'), ...new ListItemBlock().getAdminViews()];
  }

}

class ParagraphBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'paragraph';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/paragraph')];
  }

}

class UnorderedListBlock extends Block_dist_fieldsContent.Block {
  get type() {
    return 'unordered-list';
  }

  getAdminViews() {
    return [resolveView('views/editor/blocks/unordered-list'), ...new ListItemBlock().getAdminViews()];
  }

}

exports.BlockquoteBlock = BlockquoteBlock;
exports.CaptionBlock = CaptionBlock;
exports.HeadingBlock = HeadingBlock;
exports.ImageBlock = ImageBlock;
exports.ImageContainerBlock = ImageContainerBlock;
exports.LinkBlock = LinkBlock;
exports.ListItemBlock = ListItemBlock;
exports.OrderedListBlock = OrderedListBlock;
exports.ParagraphBlock = ParagraphBlock;
exports.UnorderedListBlock = UnorderedListBlock;
exports.resolveView = resolveView;
