// @keystone-ui/pill/dist/pill.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx, useTheme } from '@keystone-ui/core';
import { forwardRef } from 'react';
import { XIcon } from '@keystone-ui/icons/icons/XIcon';

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

const PillButton = (_ref) => {
  let {
    tone: toneKey,
    weight,
    onClick,
    tabIndex
  } = _ref,
      props = _objectWithoutProperties(_ref, ["tone", "weight", "onClick", "tabIndex"]);

  const {
    radii,
    spacing,
    tones,
    typography
  } = useTheme();
  const isInteractive = !!onClick;
  const tone = tones[toneKey];
  const tokens = {
    bold: {
      background: tone.fill[0],
      foreground: tone.fillForeground[0],
      focus: {
        shadow: `0 0 0 2px ${tone.focusRing}`
      },
      hover: {
        background: tone.fill[1]
      },
      active: {
        background: tone.fill[2]
      }
    },
    light: {
      background: tone.tint[0],
      foreground: tone.foreground[0],
      focus: {
        shadow: `0 0 0 2px ${tone.focusRing}`
      },
      hover: {
        foreground: tone.foreground[1],
        background: tone.tint[1]
      },
      active: {
        foreground: tone.foreground[2],
        background: tone.tint[2]
      }
    }
  }[weight];
  const baseStyles = {
    alignItems: 'center',
    appearance: 'none',
    background: 'none',
    backgroundColor: tokens.background,
    border: 0,
    color: tokens.foreground,
    display: 'flex',
    fontSize: typography.fontSize.small,
    fontWeight: typography.fontWeight.medium,
    justifyContent: 'center',
    maxWidth: '100%',
    minWidth: 1,
    outline: 0,
    padding: `${spacing.small}px ${spacing.medium}px`,
    ':first-of-type': {
      paddingRight: spacing.small,
      borderTopLeftRadius: radii.full,
      borderBottomLeftRadius: radii.full,
      marginRight: 1
    },
    ':last-of-type': {
      paddingLeft: spacing.small,
      borderTopRightRadius: radii.full,
      borderBottomRightRadius: radii.full
    },
    ':only-of-type': {
      paddingLeft: spacing.medium,
      paddingRight: spacing.medium
    }
  };
  const interactiveStyles = isInteractive ? {
    cursor: 'pointer',
    ':focus': {
      boxShadow: tokens.focus.shadow
    },
    ':hover,:focus': {
      backgroundColor: tokens.hover.background,
      color: tokens.hover.foreground
    },
    ':active': {
      backgroundColor: tokens.active.background,
      color: tokens.active.foreground
    }
  } : {};
  return jsx("button", _extends({
    css: _objectSpread2(_objectSpread2({}, baseStyles), interactiveStyles),
    onClick: onClick,
    tabIndex: !isInteractive ? -1 : tabIndex
  }, props));
};

const Pill = /*#__PURE__*/forwardRef((_ref2, ref) => {
  let {
    weight = 'bold',
    tone = 'active',
    children,
    onClick,
    onRemove
  } = _ref2,
      props = _objectWithoutProperties(_ref2, ["weight", "tone", "children", "onClick", "onRemove"]);

  console.log(onClick);
  return jsx("div", _extends({
    css: {
      display: 'flex'
    }
  }, props, {
    ref: ref
  }), jsx(PillButton, {
    weight: weight,
    tone: tone,
    onClick: onClick
  }, children), onRemove ? jsx(PillButton, {
    weight: weight,
    tone: tone,
    onClick: onRemove
  }, jsx(XIcon, {
    css: {
      height: 14,
      width: 14
    }
  })) : null);
});

export { Pill };
