// react-day-picker/lib/components/DayPicker/types/DayPickerElements.js
/**
 * List of the name of the DOM elements interface elements.
 */
export var DayPickerElements;
(function (DayPickerElements) {
    DayPickerElements["ROOT"] = "root";
    DayPickerElements["CAPTION"] = "caption";
    DayPickerElements["DAY"] = "day";
    DayPickerElements["DAYWRAPPER"] = "dayWrapper";
    // Modifiers
    DayPickerElements["SELECTED"] = "selected";
    DayPickerElements["DISABLED"] = "disabled";
    DayPickerElements["TODAY"] = "today";
    DayPickerElements["OUTSIDE_START"] = "beforemonth";
    DayPickerElements["OUTSIDE_END"] = "aftermonth";
    // Month Component
    DayPickerElements["MONTH"] = "month";
    DayPickerElements["MONTHTABLE"] = "monthTable";
    DayPickerElements["MONTHTBODY"] = "monthTbody";
    DayPickerElements["MONTHS"] = "months";
    // Head Components
    DayPickerElements["HEAD"] = "head";
    DayPickerElements["HEADROW"] = "headRow";
    DayPickerElements["HEADWEEKNUMBER"] = "headWeekNumber";
    DayPickerElements["HEADWEEKNAME"] = "headWeekName";
    /**
     * Container of [[Navigation]].
     */
    DayPickerElements["NAV"] = "nav";
    /**
     * "Previous Month" button in [[Navigation]].
     */
    DayPickerElements["NAVPREV"] = "navPrev";
    /**
     * "Next Month" button in [[Navigation]].
     */
    DayPickerElements["NAVNEXT"] = "navNext";
    // Week Component
    DayPickerElements["WEEK"] = "week";
    DayPickerElements["WEEKWEEKNUMBER"] = "weekWeeknumber";
    DayPickerElements["WEEKDAY"] = "weekDay";
    /** The weeknumber */
    DayPickerElements["WEEKNUMBER"] = "weekNumber";
})(DayPickerElements || (DayPickerElements = {}));
