// @keystone-ui/loading/dist/loading.cjs.prod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var _extends = require("@babel/runtime/helpers/extends"), _objectWithoutProperties = require("@babel/runtime/helpers/objectWithoutProperties"), core = require("@keystone-ui/core");

function _interopDefault(e) {
  return e && e.__esModule ? e : {
    default: e
  };
}

var _extends__default = _interopDefault(_extends), _objectWithoutProperties__default = _interopDefault(_objectWithoutProperties);

const loadingSizeValues = [ "large", "medium", "small" ], loadingToneValues = [ "active", "passive", "positive", "warning", "negative", "help" ], LoadingDots = _ref => {
  let {label: label, tone: toneKey, size: sizeKey = "medium"} = _ref, props = _objectWithoutProperties__default.default(_ref, [ "label", "tone", "size" ]);
  const {controlSizes: controlSizes, tones: tones} = core.useTheme(), size = controlSizes[sizeKey], tone = toneKey ? tones[toneKey] : null, color = tone ? tone.fill[0] : "currentColor";
  return core.jsx("div", _extends__default.default({
    "aria-live": "polite",
    "aria-label": label,
    css: {
      color: color,
      display: "inline-flex",
      fontSize: size.indicatorFontSize
    }
  }, props), core.jsx(Dot, {
    delay: 0
  }), core.jsx(Dot, {
    delay: 160
  }), core.jsx(Dot, {
    delay: 320
  }));
}, fadeAnimation = core.keyframes({
  "0%, 80%, 100%": {
    opacity: 0
  },
  "40%": {
    opacity: 1
  }
}), Dot = ({delay: delay}) => core.jsx("div", {
  css: {
    animation: fadeAnimation + " 1s ease-in-out infinite",
    animationDelay: delay + "ms",
    backgroundColor: "currentColor",
    borderRadius: "50%",
    display: "block",
    height: "1em",
    width: "1em",
    "&:not(:first-of-type)": {
      marginLeft: "1em"
    }
  }
});

exports.LoadingDots = LoadingDots, exports.loadingSizeValues = loadingSizeValues, 
exports.loadingToneValues = loadingToneValues;
