// @keystonejs/fields-content/views/editor/blocks/image/dist/fields-content.esm.js
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx } from '@emotion/core';
import { forwardRef, useLayoutEffect, useState, createContext, useContext } from 'react';
import _extends from '@babel/runtime/helpers/esm/extends';
import { u as useStateWithEqualityCheck } from '../../../../../dist/index-8660f797.esm.js';
import { usePopper } from 'react-popper';

const PopperRender = /*#__PURE__*/forwardRef(({
  update,
  alignment,
  isFocused,
  style,
  onAlignmentChange
}, ref) => {
  useLayoutEffect(() => {
    // The update function is `null` on first render - we want to update this after the below component is mounted to have the correct handler
    if (update) update();
  }, [alignment, update]);
  return jsx("div", {
    ref: ref,
    css: {
      display: isFocused ? 'block' : 'none',
      backgroundColor: 'black',
      padding: 8
    },
    style: style
  }, ['left', 'center', 'right'].map(align => {
    return jsx("button", {
      type: "button",
      key: align,
      onMouseDown: event => {
        // so that the image block doesn't get deselected
        event.preventDefault();
      },
      onClick: () => {
        onAlignmentChange(align);
      }
    }, align);
  }));
});
const popperModifiers = [{
  name: 'flip',
  enabled: false
}, {
  name: 'hide',
  enabled: false
}, {
  name: 'preventOverflow',
  enabled: false
}];

const Image = (_ref) => {
  let {
    alignment,
    attributes,
    isFocused,
    src,
    onAlignmentChange
  } = _ref,
      props = _objectWithoutProperties(_ref, ["alignment", "attributes", "isFocused", "src", "onAlignmentChange"]);

  const [referenceElement, setReferenceElement] = useStateWithEqualityCheck(null);
  const [popperElement, setPopperElement] = useState(null);
  const {
    styles,
    update
  } = usePopper(referenceElement, popperElement, {
    placement: 'top',
    modifiers: popperModifiers
  });
  return jsx("div", null, jsx("img", _extends({}, props, attributes, {
    src: src,
    ref: setReferenceElement,
    css: {
      width: '100%',
      outline: isFocused ? 'auto' : null
    }
  })), jsx(PopperRender, {
    update,
    alignment,
    ref: setPopperElement,
    isFocused,
    style: styles.popper,
    onAlignmentChange
  }));
};

/** @jsx jsx */
let type = 'image';
let ImageAlignmentContext = /*#__PURE__*/createContext({
  aligment: '',

  onAlignmentChange() {}

});
function Node(props) {
  let {
    data
  } = props.node;
  let {
    aligment,
    onAlignmentChange
  } = useContext(ImageAlignmentContext);
  return jsx(Image, {
    alignment: aligment,
    attributes: props.attributes,
    isFocused: props.isFocused,
    src: data.get('src'),
    onAlignmentChange: onAlignmentChange
  });
}
let getSchema = () => ({
  isVoid: true
});

export { ImageAlignmentContext, Node, getSchema, type };
