// @keystonejs/fields-content/views/editor/blocks/heading/dist/fields-content.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@babel/runtime/helpers/objectWithoutProperties');
require('@emotion/core');
var React = require('react');
require('@arch-ui/theme');
require('@babel/runtime/helpers/extends');
var views_editor_blocks_paragraph_dist_fieldsContent = require('../../paragraph/dist/fields-content.cjs.dev.js');
require('@arch-ui/color-utils');
require('@arch-ui/tooltip');
require('@arch-ui/typography');
var toolbarComponents = require('../../../../../dist/toolbar-components-5704db01.cjs.dev.js');
var utils = require('../../../../../dist/utils-5bfe488b.cjs.dev.js');

let type = 'heading';
function ToolbarElement({
  editor,
  editorState
}) {
  return /*#__PURE__*/React.createElement(toolbarComponents.ToolbarButton, {
    icon: /*#__PURE__*/React.createElement("span", {
      "aria-hidden": true
    }, "H"),
    label: "Heading",
    isActive: utils.hasBlock(editorState, type),
    onClick: () => {
      if (utils.hasBlock(editorState, type)) {
        editor.setBlocks({
          type: views_editor_blocks_paragraph_dist_fieldsContent.type
        });
      } else {
        editor.setBlocks({
          type: type
        });
      }

      editor.focus();
    }
  });
}
function Node({
  attributes,
  children
}) {
  return /*#__PURE__*/React.createElement("h2", attributes, children);
}
let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter after typing a heading,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && editor.value.blocks.every(block => block.type === type)) {
      editor.splitBlock().setBlocks(views_editor_blocks_paragraph_dist_fieldsContent.type);
      return;
    }

    next();
  }

}];

exports.Node = Node;
exports.ToolbarElement = ToolbarElement;
exports.getPlugins = getPlugins;
exports.type = type;
