// @keystone-next/fields/dist/fields.esm.js
import { Checkbox, Relationship, Text, Password, DateTimeUtc, Integer, Virtual } from '@keystonejs/fields';
import path from 'path';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import { MongoId } from '@keystonejs/fields-mongoid';
import { AutoIncrement } from '@keystonejs/fields-auto-increment';

const pkgDir = path.dirname(require.resolve('@keystone-next/fields/package.json'));
const resolveView = pathname => path.join(pkgDir, 'types', pathname);

const checkbox = (config = {}) => ({
  type: Checkbox,
  config,
  views: resolveView('checkbox/views')
});

const relationship = config => ({
  type: Relationship,
  config,
  views: resolveView('relationship/views'),
  getAdminMeta: (listKey, path, adminMetaRoot) => {
    var _config$many, _config$ui$hideCreate, _config$ui, _config$ui2, _config$ui$linkToItem, _config$ui$removeMode, _config$ui$inlineCrea, _config$ui$inlineEdit, _config$ui$inlineConn;

    const refListKey = config.ref.split('.')[0];

    if (!adminMetaRoot.listsByKey[refListKey]) {
      throw new Error(`The ref [${config.ref}] on relationship [${listKey}.${path}] is invalid`);
    }

    return _objectSpread({
      refListKey,
      many: (_config$many = config.many) !== null && _config$many !== void 0 ? _config$many : false,
      hideCreate: (_config$ui$hideCreate = (_config$ui = config.ui) === null || _config$ui === void 0 ? void 0 : _config$ui.hideCreate) !== null && _config$ui$hideCreate !== void 0 ? _config$ui$hideCreate : false
    }, ((_config$ui2 = config.ui) === null || _config$ui2 === void 0 ? void 0 : _config$ui2.displayMode) === 'cards' ? {
      displayMode: 'cards',
      cardFields: config.ui.cardFields,
      linkToItem: (_config$ui$linkToItem = config.ui.linkToItem) !== null && _config$ui$linkToItem !== void 0 ? _config$ui$linkToItem : false,
      removeMode: (_config$ui$removeMode = config.ui.removeMode) !== null && _config$ui$removeMode !== void 0 ? _config$ui$removeMode : 'disconnect',
      inlineCreate: (_config$ui$inlineCrea = config.ui.inlineCreate) !== null && _config$ui$inlineCrea !== void 0 ? _config$ui$inlineCrea : null,
      inlineEdit: (_config$ui$inlineEdit = config.ui.inlineEdit) !== null && _config$ui$inlineEdit !== void 0 ? _config$ui$inlineEdit : null,
      inlineConnect: (_config$ui$inlineConn = config.ui.inlineConnect) !== null && _config$ui$inlineConn !== void 0 ? _config$ui$inlineConn : false
    } : {
      displayMode: 'select',
      refLabelField: adminMetaRoot.listsByKey[refListKey].labelField
    });
  }
});

const text = (config = {}) => ({
  type: Text,
  config,
  views: resolveView('text/views'),
  getAdminMeta: () => {
    var _config$ui$displayMod, _config$ui;

    return {
      displayMode: (_config$ui$displayMod = (_config$ui = config.ui) === null || _config$ui === void 0 ? void 0 : _config$ui.displayMode) !== null && _config$ui$displayMod !== void 0 ? _config$ui$displayMod : 'input'
    };
  }
});

const password = (config = {}) => ({
  type: Password,
  config,
  views: resolveView('password/views'),
  getAdminMeta: () => ({
    minLength: config.minLength !== undefined ? config.minLength : 8
  })
});

const timestamp = (config = {}) => ({
  type: DateTimeUtc,
  config,
  views: resolveView('timestamp/views')
});

const integer = (config = {}) => ({
  type: Integer,
  config,
  views: resolveView('integer/views')
});

// @ts-ignore
const mongoId = config => ({
  type: MongoId,
  config,
  views: resolveView('mongoId/views')
});

// @ts-ignore
const autoIncrement = (config = {}) => ({
  type: AutoIncrement,
  config,
  views: resolveView('integer/views')
});

const select = config => ({
  type: Text,
  config,
  views: resolveView('select/views'),
  getAdminMeta: () => {
    var _config$dataType, _config$ui$displayMod, _config$ui;

    return {
      options: config.options,
      dataType: (_config$dataType = config.dataType) !== null && _config$dataType !== void 0 ? _config$dataType : 'string',
      displayMode: (_config$ui$displayMod = (_config$ui = config.ui) === null || _config$ui === void 0 ? void 0 : _config$ui.displayMode) !== null && _config$ui$displayMod !== void 0 ? _config$ui$displayMod : 'select'
    };
  }
});

// @ts-ignore
const virtual = config => ({
  type: Virtual,
  config,
  views: resolveView('virtual/views'),
  getAdminMeta: () => {
    var _config$graphQLReturn;

    return {
      graphQLReturnFragment: (_config$graphQLReturn = config.graphQLReturnFragment) !== null && _config$graphQLReturn !== void 0 ? _config$graphQLReturn : ''
    };
  }
});

export { autoIncrement, checkbox, integer, mongoId, password, relationship, select, text, timestamp, virtual };
